provider/parameter.go

```diff
@@ -459,6 +412,93 @@ func valueIsType(typ OptionType, value string) diag.Diagnostics {
 	return nil
 }
 
+func (v *Parameter) Valid() diag.Diagnostics {
+	// optionType might differ from parameter.Type. This is ok, and parameter.Type
+	// should be used for the value type, and optionType for options.
+	optionType, _, err := ValidateFormType(v.Type, len(v.Option), v.FormType)
+	if err != nil {
+		return singleDiag(diag.Diagnostic{
+			Severity:      diag.Error,
+			Summary:       "Invalid form_type for parameter",
+			Detail:        err.Error(),
+			AttributePath: cty.Path{cty.GetAttrStep{Name: "form_type"}},
+		})
+	}
+
+	optionNames := map[string]interface{}{}
+	optionValues := map[string]interface{}{}
+	if len(v.Option) > 0 {
+		for _, option := range v.Option {
+			_, exists := optionNames[option.Name]
+			if exists {
+				return singleDiag(diag.Diagnostic{
+					Severity: diag.Error,
+					Summary:  "Option names must be unique.",
+					Detail:   fmt.Sprintf("multiple options found with the same name %q", option.Name),
+				})
+			}
+			_, exists = optionValues[option.Value]
+			if exists {
+				return singleDiag(diag.Diagnostic{
+					Severity: diag.Error,
+					Summary:  "Option values must be unique.",
+					Detail:   fmt.Sprintf("multiple options found with the same value %q", option.Value),
+				})
+			}
+			diags := valueIsType(optionType, option.Value, cty.Path{})
+			if diags.HasError() {
+				return diags
+			}
+			optionValues[option.Value] = nil
+			optionNames[option.Name] = nil
+		}
+	}
+
+	if v.Default != "" && len(v.Option) > 0 {
+		if v.Type == OptionTypeListString && optionType == OptionTypeString {
+			// If the type is list(string) and optionType is string, we have
+			// to ensure all elements of the default exist as options.
+			defaultValues, diags := valueIsListString(v.Default, cty.Path{cty.GetAttrStep{Name: "default"}})
+			if diags.HasError() {
+				return diags
+			}
+
+			// missing is used to construct a more helpful error message
+			var missing []string
+			for _, defaultValue := range defaultValues {
+				_, defaultIsValid := optionValues[defaultValue]
+				if !defaultIsValid {
+					missing = append(missing, defaultValue)
+				}
+			}
+
+			if len(missing) > 0 {
+				return singleDiag(diag.Diagnostic{
+					Severity: diag.Error,
+					Summary:  "Default values must be a valid option",
+					Detail: fmt.Sprintf(
+						"default value %q is not a valid option, values %q are missing from the options",
+						v.Default, strings.Join(missing, ", "),
+					),
+					AttributePath: cty.Path{cty.GetAttrStep{Name: "default"}},
+				})
+			}
+		} else {
+			_, defaultIsValid := optionValues[v.Default]
+			if !defaultIsValid {
+				return singleDiag(diag.Diagnostic{
+					Severity:      diag.Error,
+					Summary:       "Default value must be a valid option",
+					Detail:        fmt.Sprintf("the value %q must be defined as one of options", v.Default),
+					AttributePath: cty.Path{cty.GetAttrStep{Name: "default"}},
+				})
+			}
+		}
+	}
```

Comment:
suggestion: move this block up so we can exit early and then continue the rest of the validation knowing that `len(v.Option) > 0`.

Reply:
The `if...else` blocks still exists. I can bring this `if...else` 1 scope up if I add an early exit on `default != "" || len(v.Option) == 0`. I don't think it is much of an improvement.

I am going to leave it as it was for now.